evmdis/main.go

```diff
@@ -5,29 +5,96 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
-
-	"github.com/Arachnid/evmdis"
+	"log"
+	"strconv"
+	"flag"
+	"github.com/soad003/evmdis"
 )
 
 func main() {
+	
+	withSwarmHash := flag.Bool("swarm", true, "a bool")
+	ctorMode := flag.Bool("ctor", false, "a bool")
+	logging := flag.Bool("log", false, "a bool")
+	//callsMode := flag.Bool("calls", false, "a bool")
+
+	flag.Parse()
+
+	if !*logging {
+		log.SetOutput(ioutil.Discard)
+	}
+
 	hexdata, err := ioutil.ReadAll(os.Stdin)
 	if err != nil {
 		panic(fmt.Sprintf("Could not read from stdin: %v", err))
 	}
 
-	bytecode := make([]byte, hex.DecodedLen(len(hexdata)))
+	bytecodeLength := uint64(hex.DecodedLen(len(hexdata)))
+	bytecode := make([]byte, bytecodeLength)
+
 	hex.Decode(bytecode, hexdata)
 
-	program := evmdis.NewProgram(bytecode)
-	if err := evmdis.PerformReachingAnalysis(program); err != nil {
-		panic(fmt.Sprintf("Error performing reaching analysis: %v", err))
-	}
-	evmdis.PerformReachesAnalysis(program)
-	evmdis.CreateLabels(program)
-	if err := evmdis.BuildExpressions(program); err != nil {
-		panic(fmt.Sprintf("Error building expressions: %v", err))
+	if bytecode[bytecodeLength-1] == 0x29 &&
+			bytecode[bytecodeLength-2] == 0x00 &&
+			bytecode[bytecodeLength-43] == 0xa1 &&
+			bytecode[bytecodeLength-42] == 0x65 && *withSwarmHash {
+		//swarm hash
+		bytecodeLength -= 43
+	} 
+
+	program := evmdis.NewProgram(bytecode[:bytecodeLength])
+	AnalizeProgram(program)
+
+	if *ctorMode {
+		var codeEntryPoint = FindNextCodeEntryPoint(program)
+
+		log.Printf("Entrypoint code at index %v\n", codeEntryPoint)
+
+		if codeEntryPoint == 0 {
+			panic("no code entrypoint found in ctor")
+		} else if codeEntryPoint >= bytecodeLength {
+			panic("code entrypoint outside of currently available code")
+		}
+
+		ctor := evmdis.NewProgram(bytecode[:codeEntryPoint])
+		code := evmdis.NewProgram(bytecode[codeEntryPoint:bytecodeLength])
+
+		AnalizeProgram(ctor)
+		fmt.Println("# Constructor part -------------------------")
+		PrintAnalysisResult(ctor)
+
+
+		AnalizeProgram(code)
+		fmt.Println("# Code part -------------------------")
+		PrintAnalysisResult(code)
+
+
+	} else {
+		PrintAnalysisResult(program)
+	}	
+}
+
+func FindNextCodeEntryPoint(program *evmdis.Program) uint64 {
+	var lastPos uint64 = 0
+	for _, block := range program.Blocks {
+		for _, instruction := range block.Instructions {
+			if instruction.Op == evmdis.CODECOPY {
+				var expression evmdis.Expression
+
+				instruction.Annotations.Get(&expression)
```

Comment:
I can't use InstructionExpression because every expression is saved under the key Expression (thats why i had log statments in the typemap set and get). I could have used casts probably, but don't know if that is better.